worktree.go

```diff
@@ -296,26 +297,61 @@ func (w *Worktree) ResetSparsely(opts *ResetOptions, dirs []string) error {
 	}
 
 	if opts.Mode == MixedReset || opts.Mode == MergeReset || opts.Mode == HardReset {
-		if err := w.resetIndex(t, dirs); err != nil {
+		if err := w.resetIndex(t, dirs, opts.Files); err != nil {
 			return err
 		}
 	}
 
 	if opts.Mode == MergeReset || opts.Mode == HardReset {
-		if err := w.resetWorktree(t); err != nil {
+		if err := w.resetWorktree(t, opts.Files); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
+// Restore specified files in the working tree or stage with contents from
+// a restore source. If a path is tracked but does not exist in the restore,
+// source, it will be removed to match the source.
+//
+// If Staged and Worktree are true, then the restore source will be the index.
+// If only Staged is true, then the restore source will be HEAD.
+// If only Worktree is true or neither Staged nor Worktree are true, will
+// result in ErrRestoreWorktreeeOnlyNotSupported because restoring the working
+// tree while leaving the stage untouched is not currently supported
+//
+// Restore with no files specified will return ErrNoRestorePaths
+func (w *Worktree) Restore(o *RestoreOptions) error {
+	if err := o.Validate(); err != nil {
+		return err
+	}
+
+	if o.Worktree && o.Staged {
+		// If we are doing both Worktree and Staging then it is a hard reset
+		opts := &ResetOptions{
+			Mode:  HardReset,
+			Files: o.Files,
+		}
+		return w.Reset(opts)
+	} else if o.Staged {
+		// If we are doing just staging then it is a mixed reset
+		opts := &ResetOptions{
+			Mode:  MixedReset,
+			Files: o.Files,
+		}
+		return w.Reset(opts)
+	} else {
+		return ErrRestoreWorktreeeOnlyNotSupported
+	}
```

Comment:
This `else` is redundant:

```suggestion
	}
	
	return ErrRestoreWorktreeeOnlyNotSupported
```